File: backend-spring/src/main/java/com/cronos/service/DashboardService.java
package com.cronos.service;

import com.cronos.entity.CronosOrder;
import com.cronos.repository.*;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.util.*;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class DashboardService {
    private final UserRepository userRepo;
    private final OrderRepository orderRepo;
    private final FarmRepository farmRepo;
    private final ProductRepository productRepo;
    private final CultivationCycleRepository cultivationRepo;
    private final HarvestRepository harvestRepo;
    private final ShipmentRepository shipmentRepo;

    private static final List<String> PAID_STATUSES = List.of("paid", "processing", "shipped", "delivered", "completed");

    public Map<String, Object> getAdminDashboard() {
        Map<String, Long> userCounts = userRepo.countByRole().stream().collect(Collectors.toMap(r -> (String) r[0], r -> (Long) r[1]));
        Map<String, Long> orderCounts = orderRepo.countByStatus().stream().collect(Collectors.toMap(r -> (String) r[0], r -> (Long) r[1]));
        BigDecimal totalRevenue = orderRepo.sumTotalRevenue(PAID_STATUSES);
        return Map.of("users", userCounts, "orders", orderCounts, "total_revenue", totalRevenue != null ? totalRevenue : BigDecimal.ZERO);
    }

    public Map<String, Object> getPetambakDashboard(UUID userId) {
        return Map.of(
            "total_farms", farmRepo.countByUserId(userId),
            "total_products", productRepo.countByUserId(userId),
            "total_cultivations", cultivationRepo.countByUserId(userId),
            "total_harvests", harvestRepo.countByUserId(userId),
            "total_revenue", Optional.ofNullable(orderRepo.sumRevenueBySellerIdAndStatuses(userId, PAID_STATUSES)).orElse(BigDecimal.ZERO)
        );
    }

    public Map<String, Object> getLogistikDashboard(UUID userId) {
        Map<String, Long> statusCounts = shipmentRepo.countByStatusAndCourierId(userId).stream().collect(Collectors.toMap(r -> (String) r[0], r -> (Long) r[1]));
        return Map.of("total_shipments", shipmentRepo.countByCourierId(userId), "shipment_status", statusCounts);
    }

    public Map<String, Object> getKonsumenDashboard(UUID userId) {
        var page = orderRepo.findByUserIdOrderByCreatedAtDesc(userId, PageRequest.of(0, 5));
        Map<String, Long> statusCounts = page.getContent().stream().collect(Collectors.groupingBy(CronosOrder::getStatus, Collectors.counting()));
        return Map.of("total_orders", page.getTotalElements(), "recent_orders", page.getContent(), "order_status", statusCounts);
    }
}

File: backend/services/dashboard_service.go
package services
import (
	"cronos-backend/repositories"
	"github.com/google/uuid"
)
type DashboardService struct {
	UserRepo        *repositories.UserRepository
	OrderRepo       *repositories.OrderRepository
	FarmRepo        *repositories.FarmRepository
	ProductRepo     *repositories.ProductRepository
	CultivationRepo *repositories.CultivationRepository
	HarvestRepo     *repositories.HarvestRepository
	ShipmentRepo    *repositories.ShipmentRepository
}
func NewDashboardService(
	userRepo *repositories.UserRepository,
	orderRepo *repositories.OrderRepository,
	farmRepo *repositories.FarmRepository,
	productRepo *repositories.ProductRepository,
	cultivationRepo *repositories.CultivationRepository,
	harvestRepo *repositories.HarvestRepository,
	shipmentRepo *repositories.ShipmentRepository,
) *DashboardService {
	return &DashboardService{
		UserRepo:        userRepo,
		OrderRepo:       orderRepo,
		FarmRepo:        farmRepo,
		ProductRepo:     productRepo,
		CultivationRepo: cultivationRepo,
		HarvestRepo:     harvestRepo,
		ShipmentRepo:    shipmentRepo,
	}
}
func (s *DashboardService) GetAdminDashboard() (map[string]interface{}, error) {
	userCounts, _ := s.UserRepo.CountByRole()
	orderCounts, _ := s.OrderRepo.CountByStatus()
	totalRevenue, _ := s.OrderRepo.SumTotalRevenue()
	return map[string]interface{}{
		"users":         userCounts,
		"orders":        orderCounts,
		"total_revenue": totalRevenue,
	}, nil
}
func (s *DashboardService) GetPetambakDashboard(userID uuid.UUID) (map[string]interface{}, error) {
	farmCount, _ := s.FarmRepo.CountByUserID(userID)
	productCount, _ := s.ProductRepo.CountByUserID(userID)
	cultivationCount, _ := s.CultivationRepo.CountByUserID(userID)
	harvestCount, _ := s.HarvestRepo.CountHarvestsByUserID(userID)
	revenue, _ := s.OrderRepo.SumRevenueBySellerID(userID)
	return map[string]interface{}{
		"total_farms":        farmCount,
		"total_products":     productCount,
		"total_cultivations": cultivationCount,
		"total_harvests":     harvestCount,
		"total_revenue":      revenue,
	}, nil
}
func (s *DashboardService) GetLogistikDashboard(userID uuid.UUID) (map[string]interface{}, error) {
	shipmentCount, _ := s.ShipmentRepo.CountByCourierID(userID)
	statusCounts, _ := s.ShipmentRepo.CountByStatus(userID)
	return map[string]interface{}{
		"total_shipments": shipmentCount,
		"shipment_status": statusCounts,
	}, nil
}
func (s *DashboardService) GetKonsumenDashboard(userID uuid.UUID) (map[string]interface{}, error) {
	orders, total, _ := s.OrderRepo.FindByUserID(userID, 1, 5)
	orderStatusCounts := make(map[string]int)
	for _, o := range orders {
		orderStatusCounts[o.Status]++
	}
	return map[string]interface{}{
		"total_orders":  total,
		"recent_orders": orders,
		"order_status":  orderStatusCounts,
	}, nil
}
